### PR TITLE
[OSCP]联邦学习中的安全聚合方案设计及实现 

### LINKED ISSUE #892: [heu_shamir_agg]联邦学习中的安全聚合方案设计及实现
> 此 ISSUE 为 [隐语开源共建计划（SecretFlow Open Source Contribution Plan，简称 SF OSCP）第二期](https://github.com/orgs/secretflow/discussions/741) https://github.com/secretflow/secretflow/issues/879 特别任务的子任务。
> 
> 也欢迎开发者点击[这里](https://github.com/orgs/secretflow/discussions/741)，选择你感兴趣的任务参与隐语社区共建，热烈欢迎社区开发者参与共建～

## 设计思路
by [ZacheryZ](https://github.com/ZacheryZ)
- 在聚合前调整本地梯度数据对称轴使其都变为正数，并扩大数倍使其变成整数。
- 使用PHE库中的paillier算法对本地梯度进行批量加密，每批数量与密钥长度挂钩——此步预计落在fl_base.py中。
- 重写aggregator中的加法算法，使其变为同态加——落地在aggregator中。
- 在fl_model中增加判断，如果全局权重为paillier对象，则调用解密算法再进一步分发至各个worker——此步预计落在fl_model中。

## 任务介绍
- 任务名称：【特别任务】联邦学习中的安全聚合方案设计以及实现
- 技术方向：Federated Learning
- 任务难度：挑战🌟🌟🌟
- 说明：本任务不参与🌟累计哈～

## 详细要求
安全聚合是解决联邦学习中梯度泄露问题的有效手段。然而，现有安全聚合方案仍可能存在“参与方掉线不容忍”、“聚合行为不可控”、“共谋推断不可抗”和“系统效能不实际”等问题。希望同学们能够在这个开放问题下，设计出安全，可行，巧妙的算法来解决联邦学习聚合中遇到的安全问题，效率问题，掉线问题等等。解决方法不限，可以使用 TEE，同态，多方安全计算等各种技术作为辅助。
- 安全性： 尽量少 reveal，设计出的聚合算法没有算法层面的安全问题
- 功能性：可以将实现落到 SecretFlow 的 aggregator 中，可以无缝替换其他的聚合方法。
- 收敛性：聚合结果需要保证正确性
- 代码规范：Python 代码需要使用 black+isort 进行格式化（流水线包含代码规范检查卡点）
- 提交要求：关联该 issue 并提交代码，代码提交细节请与蚂蚁导师进行沟通！！

## 能力要求
- 网空项目——安全聚合项目参与同学

## 操作说明
- 可参考 [SecretFlow 中的 aggregator](https://github.com/secretflow/secretflow/tree/main/secretflow/security/aggregation)

### PR DESCRIPTION
**Pull request by ZacheryZ**
# Fix #892

# Preamble

For Pull Request initiator, please read the following steps carefully. 请合并请求的发起人仔细阅读下列步骤。

---

If you are creating a `Pull Request` about listed topic below, please go the the `Preview` tab and click the relative link:
如果您要创建一个有关下列内容的的合并请求，请切换到 `Preview` 面板并点击下方相关的链接:

* Select [Awesome-PETs](?expand=1&template=pull_request_template_pets.md) for files under docs/awesome-pets

---

Select a proper `Pull Request` template can accelarate the review process, thanks for your cooperation.
选择一个正确的`合并请求`模版可以加速审查流程，谢谢您的配合

If your `Pull Request` topic is not amoung the list above, please just delete all prepopulated template content.
如果您的`合并请求`内容不在上述的列表里，请删除提前填入的模版内容。

---

Thanks for your contribution to SecretFlow project
感谢您对于 SecretFlow 项目的贡献

### CODE REVIEW

**Comment by ZacheryZ** [comment]
I have read the CLA Document and I hereby sign the CLA

File: secretflow/security/aggregation/experiment/phe_aggregator.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
+# Copyright 2022 Ant Group Co., Ltd.
```
Comment (zhangxingmeng) [inline secretflow/security/aggregation/experiment/phe_aggregator.py:1]: 这个聚合器名字再加一点吧，phe_aggregator太泛了，取一个特殊一点的名字吧

**Comment by zhangxingmeng** [comment]
把报错解决一下

File: secretflow/security/aggregation/experiment/ppb_aggregator.py [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+                return None
+
+    def paillier_enc_local_weights(self, local_weights):  # 加密每个参与方的weights
+        logging.info(f'local_weights{local_weights}')
+        if self._is_weights:   # weights mode
+            for index in range(len(local_weights)):
+                local_weights[index] = self.binary_encode(local_weights[index])
+                logging.info(f'binary_encode{local_weights}')
+                # self._amount_of_real_encoded_row[index] = [
+                #     len(i) for i in local_weights[index]
```
Comment (zhangxingmeng) [inline secretflow/security/aggregation/experiment/ppb_aggregator.py:89]: 废弃代码提交的时候都清理掉

**Comment by zhangxingmeng** [comment]
可以尝试pull一下master最新的代码再跑一下试试

**Comment by ZacheryZ** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by ZacheryZ** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by ZacheryZ** [comment]
recheck

**Comment by 6fj** [comment]
recheck

**Pull request closed without merging**